images/virtualization-artifact/pkg/controller/vi/internal/source/upload.go

```diff
@@ -31,28 +35,246 @@ import (
+	diskService     *service.DiskService
 }
 
 func NewUploadDataSource(
 	statService Stat,
 	uploaderService Uploader,
 	dvcrSettings *dvcr.Settings,
+	diskService *service.DiskService,
 ) *UploadDataSource {
 	return &UploadDataSource{
 		statService:     statService,
 		uploaderService: uploaderService,
 		dvcrSettings:    dvcrSettings,
+		diskService:     diskService,
 	}
 }
 
-func (ds UploadDataSource) Sync(ctx context.Context, vi *virtv2.VirtualImage) (bool, error) {
+func (ds UploadDataSource) StoreToPVC(ctx context.Context, vi *virtv2.VirtualImage) (bool, error) {
+	log, ctx := logger.GetDataSourceContext(ctx, uploadDataSource)
+
+	condition, _ := service.GetCondition(vdcondition.ReadyType, vi.Status.Conditions)
+	defer func() { service.SetCondition(condition, &vi.Status.Conditions) }()
+
+	supgen := supplements.NewGenerator(common.VIShortName, vi.Name, vi.Namespace, vi.UID)
+	pod, err := ds.uploaderService.GetPod(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+	svc, err := ds.uploaderService.GetService(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+	ing, err := ds.uploaderService.GetIngress(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+	dv, err := ds.diskService.GetDataVolume(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+	pvc, err := ds.diskService.GetPersistentVolumeClaim(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	if vi.Status.UploadCommand == "" {
+		if ing != nil && ing.Annotations[common.AnnUploadURL] != "" {
+			vi.Status.UploadCommand = fmt.Sprintf("curl %s -T example.iso", ing.Annotations[common.AnnUploadURL])
+		}
+	}
+
+	switch {
+	case isDiskProvisioningFinished(condition):
+		log.Info("Disk provisioning finished: clean up")
+
+		setPhaseConditionForFinishedImage(pvc, &condition, &vi.Status.Phase, supgen)
+
+		// Protect Ready Disk and underlying PVC.
+		err = ds.diskService.Protect(ctx, vi, nil, pvc)
+		if err != nil {
+			return false, err
+		}
+
+		// Unprotect upload time supplements to delete them later.
+		err = ds.uploaderService.Unprotect(ctx, pod, svc, ing)
+		if err != nil {
+			return false, err
+		}
+
+		err = ds.diskService.Unprotect(ctx, dv)
+		if err != nil {
+			return false, err
+		}
+
+		return CleanUpSupplements(ctx, vi, ds)
+	case common.AnyTerminating(pod, svc, ing, dv, pvc):
+		log.Info("Waiting for supplements to be terminated")
+	case pod == nil || svc == nil || ing == nil:
+		log.Info("Start import to DVCR")
+
+		envSettings := ds.getEnvSettings(vi, supgen)
+		err = ds.uploaderService.Start(ctx, envSettings, vi, supgen, datasource.NewCABundleForVMI(vi.Spec.DataSource))
+		var requeue bool
+		requeue, err = setPhaseConditionForUploaderStart(&condition, &vi.Status.Phase, err)
+		if err != nil {
+			return false, err
+		}
+
+		vi.Status.Progress = "0%"
+
+		return requeue, nil
+	case !common.IsPodComplete(pod):
+		log.Info("Provisioning to DVCR is in progress", "podPhase", pod.Status.Phase)
+
+		err = ds.statService.CheckPod(pod)
```

Comment:
There are 8 places, where this checks are used. What about to move them to `setPhaseConditionFromPodError` fn?

Reply:
Done.